### PR TITLE
copy-info: initial design doc

### PR DESCRIPTION
**Pull request by torquestomp**
# Checklist

If applicable:
- [ ] I have updated `CHANGELOG.md`
- [x] I have updated the documentation (README.md, docs/, demos/)
- [ ] I have updated the config schema (cli/src/config-schema.json)
- [ ] I have added tests to cover my changes

### CODE REVIEW

File: docs/design/copy-info.md
Comment (PhilipMetzger) [inline docs/design/copy-info.md:55]: Would it be worth it to encode this field as an enum?

File: docs/design/copy-info.md
Comment (PhilipMetzger) [inline docs/design/copy-info.md:78]: Maybe use an `Option` here, to encode the null-ability of the field as the Git backend won't use it.

**Review by PhilipMetzger — COMMENTED** [review]
Just some minor stuff, as I'm no expert. 

I do have one question though, will implicit tracking also be a future goal, where are plain `mv/cp` in a `jj` repo gets rewritten to a `jj mv/cp`? 

And a minor note on the title, if it only targets metadata it is fine. But if it also tackles #47 and #3386 then a better title would be "Copy info and Copy tracing Design" or "Copy Info and rename tracking for Jujutsu".

File: docs/design/copy-info.md
Comment (torquestomp) [reply docs/design/copy-info.md:55]: I don't think so?  It's a yes-or-no, I don't see it becoming anything else.  We already have the parent struct for holding more complex options.

Comment (PhilipMetzger) [reply docs/design/copy-info.md:55]: Yeah, I was thinking of subset fetching as on a large backend this may end up expensive.

File: docs/design/copy-info.md
Comment (martinvonz) [inline docs/design/copy-info.md:101]: We generally perform operations on commits instead and then reflect those changes in the file system. I think that would work here too, right?

I think the commands should work on any commit (i.e. take `-r` arguments). We'll probably also want `--after` flags like `hg cp/mv` has for indicating that an existing file should be considered a copy/rename (without actually changing its contents).

File: docs/design/copy-info.md
Comment (martinvonz) [inline docs/design/copy-info.md:140]: I think there are many strange and interesting merge cases to consider. For example:

```
C files: R->S=a
|
|  B files: R=b, S=b
|/
A files: R=a
```

Since commit `B` changed the contents of file `R` from `a` to `b` and commit `C` renamed file `R` to `S` (without changing its contents), the result of those two operations would be file `S` with contents `b`. Does that conflict with file `S` (with the same contents) that was added in commit `B`?

File: docs/design/copy-info.md
Comment (PhilipMetzger) [reply docs/design/copy-info.md:78]: It is a `Option` to show that `None`, is just explicitly not supported instead of using some sentinel value.  Which fits the host language better, idk how I would define that RPC api though.

File: docs/design/copy-info.md
Comment (torquestomp) [reply docs/design/copy-info.md:5]: We don't require it but we _do_ support it and that is part of the design objective, so I think it makes sense to keep it?  If we only wanted to support git-style copy info this design would be much leaner and very different.

File: docs/design/copy-info.md
Comment (torquestomp) [reply docs/design/copy-info.md:45]: Any equivalent of `hg filelog`, `hg annotate`, etc., needs this information.  I don't know if `jj` has equivalents to those yet but if it doesn't, it should.

File: docs/design/copy-info.md
Comment (torquestomp) [reply docs/design/copy-info.md:78]: On write, there's no real difference.  On read, there's a big difference: `None` means "This backend does not track copies explicitly, you need to use the other API", but `Some(empty())` means "This backend _does_ track copies explicitly and knows 100% for sure there were none in this Commit"

Comment (torquestomp) [reply docs/design/copy-info.md:78]: The distinction could also be meaningful, for instance, in a transitionary backend where some Commits were written before copy tracking was introduced, and some were written after.

File: docs/design/copy-info.md
Comment (torquestomp) [reply docs/design/copy-info.md:83]: This is mentioned a couple paragraphs down: "For backends which do not support explicit copies, it will be an error to use `--after`...".

File: docs/design/copy-info.md
Comment (torquestomp) [reply docs/design/copy-info.md:90]: I think that `jj mv A B` should still work in a Git repo by modifying the working copy as instructed.  I think it's simpler if `jj mv` is implimented to always write out the copy info in the Commit object it rewrites, and the backend silently discards it, as opposed to the client always needing to check the backend for support whenever it would write copy info.

File: docs/design/copy-info.md
Comment (torquestomp) [reply docs/design/copy-info.md:37]: Expanded this comment a fair bit.

**Comment by torquestomp** [comment]
Partial reply; I haven't added anything for merge resolution yet, need to think about it for a while.

File: docs/design/copy-info.md
Comment (martinvonz) [inline docs/design/copy-info.md:61]: nit: it feels like `roots` should come before `heads`

By the way, do you think we'll want to impose any requirement about order of copy records or do we leave it up to the backend? But if it's up to the backend, do we actually benefit from it being an iterator (instead of a vector)?

File: docs/design/copy-info.md
Comment (torquestomp) [reply docs/design/copy-info.md:45]: Clarified the comment.,

File: docs/design/copy-info.md
Comment (torquestomp) [reply docs/design/copy-info.md:5]: Updated the nomenclature here.

File: docs/design/copy-info.md
Comment (torquestomp) [reply docs/design/copy-info.md:68]: I still haven't written a section for this but I think a natural solution here is to store copy info in a `Merge<Option<T>>`, rather than a `T`, for each destination.  I think that makes it easier to approach the many corner cases we might encounter.

File: docs/design/copy-info.md
Comment (torquestomp) [reply docs/design/copy-info.md:90]: I changed this to be stricter; we actually need write->read persistence for backends that enable copy tracking, it doesn't make any sense to not have that.

File: docs/design/copy-info.md
Comment (torquestomp) [reply docs/design/copy-info.md:61]: Added a comment about ordering

File: docs/design/copy-info.md
Comment (torquestomp) [reply docs/design/copy-info.md:85]: I added a lengthy section on conflicts.

File: docs/design/copy-info.md
Comment (torquestomp) [reply docs/design/copy-info.md:172]: I think the new conflicts section covers this.

File: docs/design/copy-info.md
Comment (torquestomp) [reply docs/design/copy-info.md:77]: Reverse topological order, then

File: docs/design/copy-info.md
Comment (torquestomp) [reply docs/design/copy-info.md:96]: Hmm, so instead of a Merge<>, we could represent copy info as an ordered set (stored as a Vec<>)?  That's an interesting idea, though some backends (ours definitely) would reject multiple copies in a commit, and I'm not sure how you'd handle that for something like `jj diff` or `jj rebase`.  I chose the Merge<> struct just to represent it as a conflict that needed resolving (where `None` means "no copy source").  Maybe it should still represent that just as a Vec<> instead of a Merge<>, where len()==0 means no copy, len()==1 is resolved and len()>1 is a conflict.

Is your concern with the Merge<> type itself or do you think we should actually support copies from multiple sources in the same commit?

Comment (joyously) [reply docs/design/copy-info.md:96]: > but `CopySource` is a diff at a certain point in the DAG.

Is it, really?
> do you think we should actually support copies from multiple sources in the same commit?

Wouldn't this be like doing `jj new x y`, in which both x and y rename a file to something different, or even F1 to F to F2?

Comment (torquestomp) [reply docs/design/copy-info.md:96]: This is specifically about two copies _into_ the same destination, since copy info is keyed by destination.  Two separate copies from F -> F1 and F -> F2 can coexist without any complications.

Comment (joyously) [reply docs/design/copy-info.md:96]: OK, so it's like `jj new x y`, in which both x and y rename a file to the same name?

Comment (yuja) [reply docs/design/copy-info.md:96]: > Hmm, so instead of a Merge<>, we could represent copy info as an ordered set (stored as a Vec<>)? That's an interesting idea, though some backends (ours definitely) would reject multiple copies in a commit, and I'm not sure how you'd handle that for something like `jj diff` or `jj rebase`.

If the backend reject multiple sources, `Merge<>` (which is basically `Vec` of positive and negative terms) wouldn't work either. I thought your intention was to support that.

> I chose the Merge<> struct just to represent it as a conflict that needed resolving (where `None` means "no copy source"). Maybe it should still represent that just as a Vec<> instead of a Merge<>, where len()==0 means no copy, len()==1 is resolved and len()>1 is a conflict.
> 
> Is your concern with the Merge<> type itself or do you think we should actually support copies from multiple sources in the same commit?

My concern is what the negative `CopySource` term represents, and where it comes from for the linear squash case. Let's say we have actually squashed a merge commit `A->{B, C}`,
```
         A
        / \
{f<-b} B   C {f<-c}
        \ /
         D {f<-d}
```
we could say that the resulting commit had a copy source `f: Merge[[b, d, c]]` (if we do merge copy sources just like the other value types.) However, I have no idea what the negative term `f: d` would mean. (btw, I think `f: [b, d]` is more correct in this example.)

**Comment by PhilipMetzger** [comment]
@newren Hello Elijah, this may be of interest for you, would you mind taking a look at it?

File: docs/design/copy-info.md
Comment (torquestomp) [reply docs/design/copy-info.md:96]: I added a `CopySources` enum to represent resolved/conflicted copies without the negative elements that the `Merge` type has, I think this resolves the main issue.

> If the backend reject multiple sources
To clarify: our backend can support this (and will), but the backend's backend does not and so users will be unable to do the internal equivalent of 'git push' on such commits, which I think is fine (same as you can't push jj commits with conflicts).

**Comment by torquestomp** [comment]
> Just some minor stuff, as I'm no expert.
> 
> I do have one question though, will implicit tracking also be a future goal, where are plain `mv/cp` in a `jj` repo gets rewritten to a `jj mv/cp`?
> 
> And a minor note on the title, if it only targets metadata it is fine. But if it also tackles #47 and #3386 then a better title would be "Copy info and Copy tracing Design" or "Copy Info and rename tracking for Jujutsu".

I renamed the doc and file name to better capture the 'tracking' aspect

File: docs/design/copy-info.md
Comment (torquestomp) [reply docs/design/copy-info.md]: TIL github hides convos from you even if they're unresolved after a certain threshold.

File: docs/design/copy-info.md
Comment (torquestomp) [reply docs/design/copy-info.md]: Added a mention of `jj blame` which doesn't exist yet AFAIK

File: docs/design/copy-tracking.md
Comment (martinvonz) [inline docs/design/copy-tracking.md:175]: nit: It sounds like this is mostly generic and not related to the problem mentioned above. I think it would be clearer to describe the general rebase behavior first and then describe how we can handle the thorny problem after that.

Comment (martinvonz) [inline docs/design/copy-tracking.md:175]: This mentions how rebasing will look at both the commits between D and B and between D and C. I'm not sure it's worth explaining here, but FYI, Mercurial takes the copies from D to B and keeps only the renames from that set, then reverses the renames to get renames from B to D. It then chains that data with the copies and renames from D to C.

File: docs/design/copy-tracking.md
Comment (martinvonz) [inline docs/design/copy-tracking.md:85]: Why are the results streamed instead of being returned in a single vector? Do we have use cases where you would want to learn about some copies and then terminate the stream? Ah, is it for blame/annotate support, so that can be incrementally populated in a GUI? Could you mention that here?

File: docs/design/copy-tracking.md
Comment (martinvonz) [inline docs/design/copy-tracking.md:72]: When is this useful?

File: docs/design/copy-tracking.md
Comment (martinvonz) [inline docs/design/copy-tracking.md:262]: Ah, this is what I was asking about on the `CopySource::commit` field. Thanks for mentioning it.

File: docs/design/copy-tracking.md
Comment (torquestomp) [reply docs/design/copy-tracking.md:156]: This seems a bit slippery - what happens when you're restoring from an unrelated commit, where there is no obvious DAG to query for copy info?  (e.g., parent A, two child siblings B and C, restore from C onto B+).

File: docs/design/copy-tracking.md
Comment (torquestomp) [reply docs/design/copy-tracking.md:175]: Re-ordered this section.

File: docs/design/copy-tracking.md
Comment (torquestomp) [reply docs/design/copy-tracking.md:85]: I expect there are many cases where it's desirable for this to be a stream rather than a vector, for instance when querying the history of a custom backend file with many many many revisions and running some sort of search routine on it, or intersection, etc.  Added an explicit note for blame/annotate

File: docs/design/copy-tracking.md
Comment (torquestomp) [reply docs/design/copy-tracking.md:72]: I think it's not relevant to git, but it seems like a pretty significant performance knob for not-Git?  That said I'm not actually sure, other than as a flag to `jj blame` in the future i.e. if the user doesn't want to follow copies.

I was imagining cases where you only want 'B' and not 'A' but I can't actually pin them down, it seems like you'd always want 'A' since you're specifying a range limit on the query.

**Review by martinvonz — COMMENTED** [review]
I think this looks good enough to start implementing. @fowles has already started working on it. Can we get an approval from a non-Googler? Thanks

**Review by PhilipMetzger — APPROVED** [review]
LG

File: docs/design/copy-tracking.md
Comment (torquestomp) [reply docs/design/copy-tracking.md]: If we don't support this upstream then any extension that does support it would need to rewrite 'diff', 'rebase', everything etc. that reads the copy source contents.  I'll leave this as-is for now, can iterate further in other PRs.

**Pull request merged**
